### PR TITLE
Remove option to enable direct buffer pooling

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit removes the option to change the netty system properties to
reenable the direct buffer pooling. It also removes the need for us to
disable the buffer pooling in the system properties file. Instead, we
programmatically craete an allocator that is used by our networking
layer.

This commit does introduce an Elasticsearch property which allows the
user to fallback on the netty default allocator. If they choose this
option, they can configure the default allocator how they wish using the
standard netty properties.

### CODE REVIEW

**Comment by tbrooks8** [comment]
When we disable direct buffer pooling in 7.4, we agreed that once it was determined to be stable we could remove the fallback mode. This commit moves to a world where we programmatically assign the buffer allocator and do not use the default netty one. 

This commit does add a system property to allow expert users to fallback on the netty allocator (and therefore configure it how they like).

We still need to discuss this. The name of the system property is tentative. And this has some friction with the "unpooled" system property which we currently apply. We should perhaps move to a world where JvmErgonomics just tells us if we should use the unpooled allocator. However, then this has some tension with testing (we could add an Elasticsearch system property for the unpooled allocator the tests apply `es.allocator.unpooled` or something).

**Comment by original-brownbear** [comment]
> allows the user to fallback on the netty default allocator.

I wonder if we should even allow this? I'm having a hard time seeing any situation where this would be helpful. On the other hand, if we keep this option around we have to continue to support/test it.
I think I'd rather just stick with `CopyBytesSocketChannel` since that seems to work in a safe and stable way now and keep it simple unless there's a good example for a situation where manually overriding the details of the allocator would be helpful?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java [hunk 1]
```diff
@@ -216,14 +208,14 @@ private void createServerBootstrap(ProfileSettings profileSettings, NioEventLoop
...
-        // per-event-loop thread to be used for IO operations.
-        if (ByteBufAllocator.DEFAULT.isDirectBufferPooled()) {
-            serverBootstrap.channel(NioServerSocketChannel.class);
-        } else {
-            serverBootstrap.channel(CopyBytesServerSocketChannel.class);
-        }
+        // CopyBytesServerSocketChannel which will create child channels of type CopyBytesSocketChannel.
+        // CopyBytesSocketChannel pool a single direct buffer per-event-loop thread to be used for IO
+        // operations.
+        serverBootstrap.channel(CopyBytesServerSocketChannel.class);
```
Comment (ywelsch) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java]: Why can we unconditionally set this now? Shouldn't we check with the `NettyAllocator.ALLOCATOR`?

File: distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmErgonomics.java [hunk 2]
```diff
@@ -56,11 +56,11 @@ private JvmErgonomics() {
         final Map<String, Optional<String>> finalJvmOptions = finalJvmOptions(userDefinedJvmOptions);
         final long heapSize = extractHeapSize(finalJvmOptions);
         final Map<String, String> systemProperties = extractSystemProperties(userDefinedJvmOptions);
-        if (systemProperties.containsKey("io.netty.allocator.type") == false) {
+        if (systemProperties.containsKey("es.unsafe.use_unpooled_allocator") == false) {
```
Comment (ywelsch) [inline distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmErgonomics.java]: this setting does not need the unsafe prefix, that's only relevant for `es.unsafe.use_netty_default_allocator`

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left some minor comments, o.w. looking good

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyAllocator.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        public ByteBuf directBuffer(int initialCapacity) {
+            // TODO: Currently the Netty SslHandler requests direct ByteBufs even when interacting with the
+            //  JDK SSLEngine. This will be fixed in a future version of Netty. For now, return a heap
+            //  ByteBuf. After a Netty upgrade, return to throwing UnsupportedOperationException
+            return heapBuffer(initialCapacity);
+        }
+
+        @Override
+        public ByteBuf directBuffer(int initialCapacity, int maxCapacity) {
+            // TODO: Currently the Netty SslHandler requests direct ByteBufs even when interacting with the
```
Comment (original-brownbear) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyAllocator.java]: NIT: Maybe link the relevant Netty issue in the TODO?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyAllocator.java [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        }
+
+        @Override
+        public CompositeByteBuf compositeDirectBuffer(int maxNumComponents) {
+            throw new UnsupportedOperationException("Direct buffers not supported.");
+        }
+
+        @Override
+        public boolean isDirectBufferPooled() {
+            return delegate.isDirectBufferPooled();
```
Comment (original-brownbear) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyAllocator.java]: `return false;` ?

File: modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpClient.java [hunk 5]
```diff
@@ -84,7 +85,10 @@
     private final Bootstrap clientBootstrap;
 
     Netty4HttpClient() {
-        clientBootstrap = new Bootstrap().channel(NioSocketChannel.class).group(new NioEventLoopGroup());
+        clientBootstrap = new Bootstrap()
```
Comment (original-brownbear) [inline modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpClient.java]: Could we maybe not change this or make it use a single threaded `NioEventLoopGroup` until we've successfully dealt with https://github.com/elastic/elasticsearch/issues/46803 (just reopened, the fact that every thread we touch is using 1MB of direct memory, weakly-referenced as a `ThreadLocal` with the copy bytes channel is causing trouble in CI in some tests and this seems like a spot where we'd add more trouble potentially) in full?

**Review by original-brownbear — COMMENTED** [review]
Just one comment on the test http client change, otherwise looks good :)

**Review by ywelsch — CHANGES_REQUESTED** [review]

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyAllocator.java [hunk 6]
```diff
@@ -202,7 +202,7 @@ public CompositeByteBuf compositeDirectBuffer(int maxNumComponents) {
 
         @Override
         public boolean isDirectBufferPooled() {
-            return delegate.isDirectBufferPooled();
+            return false;
```
Comment (ywelsch) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyAllocator.java]: perhaps `assert delegate.isDirectBufferPooled() == false`?

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**